Close inactive connections and requests

WyriHaximus opened the pull request:
This builds on top of #405 and further builds out #423 by also
close connections with inactive requests.

Closes: #423

mmoreram commented:
@WyriHaximus is there any chance you can merge this PR? Seems to be tested, and increases so much the memory performance :)

WyriHaximus commented:
> @WyriHaximus Thanks for the update!
> 
> Can you provide a quick update here, what's your plan for this feature with regards to #425? We've briefly discussed (off channel) that this should eventually send a `408 Request Timeout` response for the initial connection but not for keep-alive connections that send no following request. I agree breaking this up into smaller PRs can be useful, but it looks like this might also work in a single PR, so I'll leave this up to you. Let me know if there's anything I can help with and/or if you want me to take over from here +1

@clue Right, been thinking about this, and doesn't it make sense to do this with all requests in flight, instead of only on the first?

WyriHaximus commented:
> @WyriHaximus is there any chance you can merge this PR? Seems to be tested, and increases so much the memory performance :)

@mmoreram There are a few details we need to polish before this PR is done and it can be merged. Like the 408 responses for when an inflight request times out. And ensuring we don't cut off websocket connections with this.

WyriHaximus commented:
> > @WyriHaximus Thanks for the update!
> > Can you provide a quick update here, what's your plan for this feature with regards to #425? We've briefly discussed (off channel) that this should eventually send a `408 Request Timeout` response for the initial connection but not for keep-alive connections that send no following request. I agree breaking this up into smaller PRs can be useful, but it looks like this might also work in a single PR, so I'll leave this up to you. Let me know if there's anything I can help with and/or if you want me to take over from here +1
> 
> @clue Right, been thinking about this, and doesn't it make sense to do this with all requests in flight, instead of only on the first?

Just made this change, the code is cleaner as well. I Will check the relevant RFC's later today, but until that my position is that any request in a keep alive connection can timeout for any number of reasons.

clue commented:
> > > @WyriHaximus Thanks for the update!
> > > Can you provide a quick update here, what's your plan for this feature with regards to #425? We've briefly discussed (off channel) that this should eventually send a `408 Request Timeout` response for the initial connection but not for keep-alive connections that send no following request. I agree breaking this up into smaller PRs can be useful, but it looks like this might also work in a single PR, so I'll leave this up to you. Let me know if there's anything I can help with and/or if you want me to take over from here +1
> > 
> > 
> > @clue Right, been thinking about this, and doesn't it make sense to do this with all requests in flight, instead of only on the first?
> 
> Just made this change, the code is cleaner as well. I Will check the relevant RFC's later today, but until that my position is that any request in a keep alive connection can timeout for any number of reasons.

Did take a look at the relevant RFCs (https://datatracker.ietf.org/doc/html/rfc7230#section-6 and https://datatracker.ietf.org/doc/html/rfc7231#section-6.5.7 come to mind) and it looks like this is indeed under-specified behavior. MDN mentions "Note: some servers merely shut down the connection without sending this message." (https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/408) and this actually seems to match my observations.

* If you open a connection github.com:80 or google.com:80 and don't send a complete request, the connection will be closed with no message after a timeout. Same if you send a request using keep-alive and then wait for the timeout. Same if you send a request using keep-alive and then send an incomplete request header.

* If you open a connection to Apache and don't send any data, the connection will be closed with no message after a timeout. If you open a connection and send an incomplete request header, the connection will be closed with a message after a timeout. In either case, a 408 will be logged. If you send a request using keep-alive and then wait for the timeout, the connection will be closed with no message after a timeout and no log message. If you send a request using keep-alive and then send an incomplete request header, the connection will be closed with an message after a timeout and a log message.

* If you open a connection to nginx, it will close the connection with no message after after a timeout in either case. Even though I'm using default settings (http://nginx.org/en/docs/http/ngx_http_core_module.html#client_header_timeout), I've never seen an 408 response in my tests.

* Note there's also https://www.haproxy.com/blog/haproxy-and-http-errors-408-in-chrome/ which suggests sending a 408 used to be(?) a problem in chrome to its inherent race condition. See also https://stackoverflow.com/questions/42631273/how-do-browsers-handle-http-keepalive-race-condition for some background.

Neither behavior appears to be against specs, so I wonder what behavior we should use here.

WyriHaximus commented on src/Io/StreamingServer.php:
```diff
@@ -96,18 +98,20 @@ final class StreamingServer extends EventEmitter
 
         $this->callback = $requestHandler;
-        $this->parser = new RequestHeaderParser();
+        $this->parser = new RequestHeaderParser($this->loop, $this->idleConnectionTimeout);
```
As in also make it possible to inject a different header parser? We could do that so we can have a C based extension for parser to speed it up?

clue replied:
KISS: The `StreamingServer` is marked `@internal`, as is the `RequestHeaderParser`. This means we can easily inject the `RequestHeaderParser` to the `StreamingServer` inside the `Server` class. Constructor may have to be changed, but doesn't affect BC, because the public `Server` interface is not affected (at least not as part of this change).

WyriHaximus replied:
Yes, I'm fully aware, I'm also thinking ahead. But maybe, if we want to go there, we should ship that C based extension to speed it up

clue replied:
I'm not sure this is currently our hot path, but if we have evidence an optional extension will speed things up, I'm all for it 👍 I think we're on the same page here and should probably leave this up for a follow-up PR?

In either case I think your changes are a welcome improvement as is and injecting the `RequestHeaderParser` into the `StreamingServer` can help simplify the dependency injection both now and longer-term 👍

WyriHaximus replied:
> I'm not sure this is currently our hot path, but if we have evidence an optional extension will speed things up, I'm all for it 👍 I think we're on the same page here and should probably leave this up for a follow-up PR?

Of course, that would be a follow-up PR. And we have tons of other things to do.

> In either case I think your changes are a welcome improvement as is and injecting the `RequestHeaderParser` into the `StreamingServer` can help simplify the dependency injection both now and longer-term 👍

Sounds good! 🚢 🇮🇹 !

(Will fix that thing below this now.)

clue replied:
It's my understanding the `HttpServer` should be simplified to call something like `new StreamingServer($middleware, new RequestHeaderParser($loop, $idleConnectionTimeout))`.

WyriHaximus replied:
It took some work, but I think it's worth the refactoring :heart:

WyriHaximus replied:
Took some working on this, but ended up with `new StreamingServer($middleware, $idleConnectionTimeout)` because the connection/requests timing out is handled in the `StreamingServer`.

WyriHaximus commented on src/Io/RequestHeaderParser.php:
```diff
@@ -24,12 +25,44 @@ class RequestHeaderParser extends EventEmitter
-        $conn->on('data', $fn = function ($data) use (&$buffer, &$fn, $conn, $maxSize, $that) {
+        $conn->on('data', $fn = function ($data) use (&$buffer, &$fn, $conn, $maxSize, $that, $loop, &$timer, $idleConnectionTimeout) {
+            $loop->cancelTimer($timer);
+            $timer = $loop->addTimer($idleConnectionTimeout, function () use ($conn) {
+                $conn->end("HTTP/1.1 " . Response::STATUS_REQUEST_TIMEOUT . " Request Timed Out\r\n");
+            });
```
I think it should

WyriHaximus commented:
> Neither behavior appears to be against specs, so I wonder what behavior we should use here.

@clue For now, merge this as is and I'll do a follow up PR to make the behavior configurable.

clue commented on src/Io/RequestHeaderParser.php:
```diff
@@ -24,12 +25,44 @@ class RequestHeaderParser extends EventEmitter
+        $timer = $loop->addTimer($idleConnectionTimeout, function () use ($conn) {
+            $conn->close();
+        });
+        $conn->on('close', function () use ($loop, $timer) {
```
`$timer` needs to be passed by reference here, as the `data` event could reschedule a new timer at any time.

clue commented on src/Io/RequestHeaderParser.php:
```diff
@@ -24,12 +25,44 @@ class RequestHeaderParser extends EventEmitter
+        $conn->on('data', $fn = function ($data) use (&$buffer, &$fn, $conn, $maxSize, $that, $loop, &$timer, $idleConnectionTimeout) {
+            $loop->cancelTimer($timer);
+            $timer = $loop->addTimer($idleConnectionTimeout, function () use ($conn) {
+                $conn->end("HTTP/1.1 " . Response::STATUS_REQUEST_TIMEOUT . " Request Timed Out\r\n\r\n");
```
Based on https://github.com/reactphp/http/pull/425#issuecomment-1030671101, I would argue it's safer to just close the connection without sending any response. There's an inherent race condition here, simply closing the connection at least allows the client to identify this problematic behavior.

WyriHaximus replied:
:+1:

WyriHaximus replied:
Interestingly enough, using `close` over an empty `end` blows up. Looking into why

WyriHaximus replied:
Ok found out why, and it looks like we have some tests leakage somewhere

clue commented on src/Middleware/InactiveConnectionTimeoutMiddleware.php:
```diff
@@ -0,0 +1,58 @@
+ */
+final class InactiveConnectionTimeoutMiddleware
+{
+    const DEFAULT_TIMEOUT = 60;
```
Should this be `@internal` so we can change the name/value in the future?

WyriHaximus replied:
Yes it should

clue commented on src/Io/StreamingServer.php:
```diff
@@ -134,7 +138,13 @@ public function __construct(LoopInterface $loop, $requestHandler)
+    /** @internal */
+    public function handle(ConnectionInterface $conn)
+    {
+        $this->parser->handle($conn);
```
Leftover?

clue commented on src/Io/StreamingServer.php:
```diff
@@ -352,7 +362,7 @@ public function handleResponse(ConnectionInterface $connection, ServerRequestInt
             // either wait for next request over persistent connection or end connection
             if ($persist) {
-                $this->parser->handle($connection);
+                $this->handle($connection);
```
Leftover?

clue commented on src/Io/StreamingServer.php:
```diff
@@ -373,10 +383,10 @@ public function handleResponse(ConnectionInterface $connection, ServerRequestInt
+            $body->on('end', function () use ($connection, $that, $body) {
                 $connection->removeListener('close', array($body, 'close'));
-                $parser->handle($connection);
+                $that->handle($connection);
```
Leftover?

clue commented on src/Io/StreamingServer.php:
```diff
@@ -85,6 +86,7 @@ final class StreamingServer extends EventEmitter
     private $callback;
     private $parser;
     private $loop;
+    private $idleConnectionTimeout;
```
Leftover?

clue commented on src/Io/StreamingServer.php:
```diff
@@ -96,18 +98,20 @@ final class StreamingServer extends EventEmitter
         }
 
         $this->loop = $loop;
+        $this->idleConnectionTimeout = $idleConnectTimeout;
```
Leftover?

clue commented on tests/HttpServerTest.php:
```diff
@@ -257,6 +267,18 @@ function (ServerRequestInterface $request) use (&$streaming) {
+//        $this->socket->emit('connection', array($this->connection));
+//
+//        Loop::run();
+//    }
```
Leftover?

WyriHaximus replied:
It was supposed to be a test. So will uncomment and test it.

clue commented on src/Middleware/InactiveConnectionTimeoutMiddleware.php:
```diff
@@ -0,0 +1,61 @@
+    /**
+     * @return float
+     */
+    public function getTimeout()
```
Should this be `@internal`?

WyriHaximus replied:
Yes.

WyriHaximus replied:
Yes.

clue commented on tests/Io/RequestHeaderParserTest.php:
```diff
@@ -831,6 +833,35 @@ public function testQueryParmetersWillBeSet()
+        Loop::addTimer(0.2, function () {
+            Loop::stop();
+        });
+        Loop::run();
```
This seems quite fragile, should we mock the loop here instead?

WyriHaximus replied:
Yes, working on it

WyriHaximus replied:
Done

clue commented on tests/Io/RequestHeaderParserTest.php:
```diff
@@ -831,6 +833,35 @@ public function testQueryParmetersWillBeSet()
+        $connection = $this->getMockBuilder('React\Socket\Connection')->disableOriginalConstructor()->setMethods(array('close'))->getMock();
+        $connection->expects($this->never())->method('close');
+
+        $parser = new RequestHeaderParser(Loop::get(), new Clock(Loop::get()), 0.1);
```
This seems quite fragile, should we mock the loop here instead?

WyriHaximus replied:
Done

clue commented on src/Io/RequestHeaderParser.php:
```diff
@@ -24,20 +25,48 @@ class RequestHeaderParser extends EventEmitter
+            $loop->cancelTimer($timer);
+            $timer = $loop->addTimer($idleConnectionTimeout, function () use ($conn) {
+                $conn->close();
+            });
```
Not sure how relevant this is, but considering this is on the hot path I'd like to optimize for the happy path here: Remove timer on `data` and only restart if request is incomplete (unlikely). At the moment, this would cancel + add + cancel.

WyriHaximus replied:
Done

clue commented on src/Io/StreamingServer.php:
```diff
@@ -96,19 +97,19 @@ final class StreamingServer extends EventEmitter
      * @see self::listen()
      */
-    public function __construct(LoopInterface $loop, $requestHandler)
+    public function __construct($requestHandler, RequestHeaderParser $parser, Clock $clock)
```
This change makes the diff quite massive, I wonder if that's something we can reasonably avoid?

WyriHaximus replied:
What do you have in mind? Regardless of what we do it will be a massive diff. What I could do is file another PR updating the tests moving some things around that will make the diff in this PR a lot less.

WyriHaximus replied:
Just filed https://github.com/reactphp/http/pull/469 preparing to keep the change set smaller here,

clue commented on src/Io/RequestHeaderParser.php:
```diff
@@ -24,20 +25,48 @@ class RequestHeaderParser extends EventEmitter
+        $loop = $this->loop;
+        $idleConnectionTimeout = $this->idleConnectionTimeout;
+        $timer = $loop->addTimer($idleConnectionTimeout, function () use ($conn) {
+            $conn->close();
```
Should this trigger an `error` event similar to how an overflow would trigger?

WyriHaximus replied:
Yes, working on it

WyriHaximus replied:
Done

WyriHaximus committed:
Typo

WyriHaximus committed:
Close inactive connections and requests

This new middleware introduces a timeout of closing inactive
connections between requests after a configured amount of seconds.

This builds on top of #405 and partially on #422
